src/gnn_reco/components/utils.py

```diff
@@ -0,0 +1,266 @@
+                else:
+                    pbar.set_description('epoch: %s || loss: %s'%(epoch, acc_loss.item()/iteration))
+                iteration +=1
+            if self._early_stopping_method.step(validation_loss,model):
+                print('EARLY STOPPING: %s'%epoch)
+                break
+        return model
+            
+    def _validate(self,model):
+        acc_loss = torch.tensor([0],dtype = float).to(self.device)
+        model.eval()
+        for batch_of_graphs in self.validation_dataloader:
+            batch_of_graphs.to(self.device)
+            with torch.no_grad():                                                                                        
+                out                             = model(batch_of_graphs)   
+                loss                            = self.loss_func(out, batch_of_graphs, self.target)                             
+                acc_loss += loss
+        return acc_loss/len(self.validation_dataloader)
+    
+    def _load_best_parameters(self,model):
+        return model.load_state_dict(self._early_stopping_method.GetBestParams())
+         
+
+class Predictor(object):
+    def __init__(self, dataloader, target, device, output_column_names, post_processing_method = None):
+        self.dataloader = dataloader
+        self.target = target
+        self.output_column_names = output_column_names
+        self.device = device
+        self.post_processing_method = post_processing_method
+    def __call__(self, model):
+        self.model = model
+        self.model.eval()
+        self.model.predict = True
+        if self.post_processing_method == None:
+            return self._predict()
+        else:
+            return self.post_processing_method(self._predict(),self.target)
+
+    def _predict(self):
+        predictions = []
+        event_nos   = []
+        target      = []
+        with torch.no_grad():
+            for batch_of_graphs in tqdm(self.dataloader, unit = 'batches'):
+                batch_of_graphs.to(self.device)
+                target.extend(batch_of_graphs[self.target].detach().cpu().numpy())
+                predictions.extend(self.model(batch_of_graphs).detach().cpu().numpy())
+                event_nos.extend(batch_of_graphs['event_no'].detach().cpu().numpy())
+        out = pd.DataFrame(data = predictions, columns = self.output_column_names)
+        out['event_no'] = event_nos
+        out[self.target] = target
+        return out
+
+def vonmises_sine_cosine_post_processing(predictions,target):
+    predictions[target + '_pred'] = np.arctan2(predictions['sine'], predictions['cosine'])
+    predictions['k'] = abs(predictions['k'])
+    predictions['sigma'] = 1/predictions['k']
+    return predictions        
+
+
+def make_train_validation_dataloader(db, selection, pulsemap, batch_size, FEATURES, TRUTH, num_workers):
+    training_selection, validation_selection = train_test_split(selection, test_size=0.33, random_state=42)
```

Comment:
Not a change request, just a note: We should decide where in the pipeline, from I3 files to model inference, we want to define train-and-test-splits such that these become unambiguous and transferrable between models/runs/projects. I have always used modulo of event numbers as a reproducible way of making these splits, but am interested to hear your thoughts, @RasmusOrsoe, @mhaminh, and @philippeller.